### ISSUE TITLE
why dose the loss of photo_detail_mrf is nan?

### ISSUE THREAD

**Issue by kingo233** [YadiraF/DECA#175]
Hi! thank you for your excellent work.
I have finished pretrain and corase shape train,and the model's ablibity to extract pose,light,tex is good.
But when I started trainning detail,I got photo_detail_mrf is nan .
How can I solve it?

**Comment by zydmu123** [comment]
@kingo233 Hello, I'm also trying to reproduce the training process. However, after the coarse training, I couldn't get the right code['light'] while the other latent codes are well learned, I wonder can you give me some suggestions?

**Comment by kingo233** [comment]
@zydmu123 Hi! Please give me more infomation about this problem.Can you show me your training image ?

**Comment by zydmu123** [comment]
@kingo233 Thanks, I have fixed the light, just by switching the below codes:

**Comment by zydmu123** [comment]
@kingo233 Thanks, I have fixed the light, just by switching the below codes:
`# ops = self.render(verts, trans_verts, albedo, codedict['light'])
    ops = self.render(verts, trans_verts, albedo, h=h, w=w, background=background)`
and my coarse training results for now are as follows：
![00039350](https://user-images.githubusercontent.com/111970347/227425556-7f18728c-e466-4fcf-bfd4-b33a5ae0fb76.jpg)

However, a similar problem as yours seems to arise during my detail training...

**Comment by kingo233** [comment]
@zydmu123 Hi! I haven't solved this problem. Can you give your contact infomation? Want to communicate with you more directly

**Comment by zydmu123** [comment]
@kingo233 here is my e-mail: bh_landmarks3d@foxmail.com

**Comment by kingo233** [comment]
@zydmu123 Hey,bro. Please check your email.Our team member have sent you an email talking about ours situation.

**Comment by xuduo18311199384** [comment]
@kingo233 @zydmu123 Hello, I train coarse well. But detail result very bad. Have you trained detail well? My wechat: xuduo1029

![orig_11067_inputs](https://github.com/yfeng95/DECA/assets/30895764/34299808-d06b-4f68-a07c-79c5c1e65e7c)
![2023-07-28 14-18-22 的屏幕截图](https://github.com/yfeng95/DECA/assets/30895764/00aeeb9a-ce5d-417d-a16c-d1a82fa5ce27)
![2023-08-04 17-29-00 的屏幕截图](https://github.com/yfeng95/DECA/assets/30895764/32269754-1e74-40f0-84de-ed24123c98e6)

**Comment by zek1th** [comment]
> Hi! thank you for your excellent work. I have finished pretrain and corase shape train,and the model's ablibity to extract pose,light,tex is good. But when I started trainning detail,I got photo_detail_mrf is nan . How can I solve it?
我也遇到这个问题了，能问下你是怎么解决的吗？还有，文中说在训练细节模型时，粗模型是固定的，这代表我要将训练粗模型时的loss都置为0吗？还是需要什么别的操作啊。如能回复，感激不尽

**Comment by kingo233** [comment]
> > 我也遇到这个问题了，能问下你是怎么解决的吗？还有，文中说在训练细节模型时，粗模型是固定的，这代表我要将训练粗模型时的loss都置为0吗？还是需要什么别的操作啊。如能回复，感激不尽

这代表你的训练配置文件要把is_detail设置为true，训练时代码会走到另外一个分支，粗模型部分不会得到更新

**Comment by zek1th** [comment]
> > > 我也遇到这个问题了，能问下你是怎么解决的吗？还有，文中说在训练细节模型时，粗模型是固定的，这代表我要将训练粗模型时的loss都置为0吗？还是需要什么别的操作啊。如能回复，感激不尽
> 
> 这代表你的训练配置文件要把is_detail设置为true，训练时代码会走到另外一个分支，粗模型部分不会得到更新

原来如此，感谢感谢，我已经设置cfg.train.train_detail = true。
但是另一个我跟你之前存在的问题一样，detail阶段的 photo_detail_mrf 值为 nan，这个你是怎么解决的呢？

**Comment by kingo233** [comment]
> 原来如此，感谢感谢，我已经设置cfg.train.train_detail = true。 但是另一个我跟你之前存在的问题一样，detail阶段的 photo_detail_mrf 值为 nan，这个你是怎么解决的呢？

这个我也一直没解决，抱歉

**Comment by zek1th** [comment]
> > 原来如此，感谢感谢，我已经设置cfg.train.train_detail = true。 但是另一个我跟你之前存在的问题一样，detail阶段的 photo_detail_mrf 值为 nan，这个你是怎么解决的呢？
> 
> 这个我也一直没解决，抱歉

好吧，谢谢了:>

**Comment by BYF-ai** [comment]
Many of the losses are NaN during the fine-tuning phase. What should I do

**Comment by BYF-ai** [comment]
Many of the losses are NaN during the detail phase. What should I do

**Comment by wrlingotion** [comment]
Did any of you manage to solve the photo_detail_mrf loss issue?

**Comment by kingo233** [comment]
> Did any of you manage to solve the photo_detail_mrf loss issue?

Sorry not yet

**Comment by zek1th** [comment]
> Did any of you manage to solve the photo_detail_mrf loss issue?

Acorrding to the MRF-loss paper, I think its because the relative distance's caculating is not right in any situation. So I set the limit of the "cosine_dist_i" (in ’mrf_loss‘ function) as [-1,1] like the picture. It works but I'm not sure.

![Image](https://github.com/user-attachments/assets/f137d995-8015-49d9-8b9f-4add77ece81b)

**Comment by papekoja** [comment]
@zek1th hmm, alright we also tried that but with a higher value to debug. I guess that will force it to work. I wonder if it trains properly? Since the core issue must still be there.
